src/script/builder/tir/block_frame.cc

```diff
@@ -39,24 +42,114 @@ BlockFrame Block_(String name) {
   n->annotations.clear();
   n->iter_values.clear();
   n->predicate = NullOpt;
+  n->no_realize = no_realize;
   return BlockFrame(n);
 }
 
 void BlockFrameNode::ExitWithScope() {
   using namespace tvm::tir;
   TIRFrameNode::ExitWithScope();
-  AddToParent(BlockRealize(iter_values,  //
-                           predicate.value_or(Bool(true)),
-                           Block(iter_vars,      //
-                                 reads, writes,  //
-                                 name,           //
-                                 AsStmt(stmts),  //
-                                 init,           //
-                                 alloc_buffers,  //
-                                 match_buffers,  //
-                                 annotations)));
+  Block block = Block(iter_vars, reads, writes, name, AsStmt(stmts), init, alloc_buffers,
+                      match_buffers, annotations);
+  if (no_realize && iter_values.empty() && !predicate.defined()) {
+    AddToParent(block);
```

Comment:
Ah I think we should check and error out if `iter_values` or `predicate` is defined.

```suggestion
  if (no_realize) {
    CHECK(!iter_values.empty()) << "...";
    CHECK(predicate.defined()) << "...";
```